DRILL-6284: Add operator metrics for batch sizing for flatten

ppadma opened the pull request:
@kkhatua please review.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java:
```diff
@@ -99,6 +100,22 @@ private void clear() {
   }
 
+  public enum Metric implements MetricDef {
+    NUM_INCOMING_BATCHES,
```
Strangely, these enum names show up as labels in the web UI. Maybe `INCOMING_BATCH_COUNT`? The `AVG_` and `TOTAL_` names seem readable enough...

paul-rogers replied:
Also, maybe `INPUT_BATCH_COUNT` to be parallel with `AVG_INPUT_BATCH_SIZE`?

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java:
```diff
@@ -99,6 +100,22 @@ private void clear() {
+  public enum Metric implements MetricDef {
+    NUM_INCOMING_BATCHES,
+    AVG_INPUT_BATCH_SIZE,
+    AVG_INPUT_ROW_WIDTH,
```
`AVG_INPUT_ROW_WIDTH` to be parallel with `AVG_OUTPUT_ROW_WIDTH`?

ppadma replied:
@paul-rogers Paul, I did not understand what you mean by parallel here and below. Do you mean they should be adjacent columns in the web UI ?

paul-rogers replied:
Parallel here simply means the same name: "INCOMING"/"INPUT" --> "INPUT", "OUTGOING"/"OUTPUT" --> "OUTPUT". Same name in each term.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java:
```diff
@@ -99,6 +100,22 @@ private void clear() {
 
+  public enum Metric implements MetricDef {
+    NUM_INCOMING_BATCHES,
+    AVG_INPUT_BATCH_SIZE,
```
Not clear what units `SIZE` is in. Maybe `AVG_INPUT_BATCH_BYTES` (if it is bytes)?

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java:
```diff
@@ -99,6 +100,22 @@ private void clear() {
+    AVG_INPUT_BATCH_SIZE,
+    AVG_INPUT_ROW_WIDTH,
+    TOTAL_INPUT_RECORDS,
+    NUM_OUTGOING_BATCHES,
```
`OUTPUT_BATCH_COUNT` as per comments above?

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java:
```diff
@@ -463,4 +488,20 @@ protected boolean setupNewSchema() throws SchemaChangeException {
+  }
+
+  @Override
+  public void close() {
```
Is `super.close()` needed?

paul-rogers replied:
Seems resolved.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/AbstractRecordBatchMemoryManager.java:
```diff
@@ -29,6 +29,50 @@
+  }
+
+  public long getAvgOutputRowWidth() {
+    return RecordBatchSizer.safeDivide(sumOutputRowWidths, numOutgoingBatches);
```
Not sure if it really matters, but this calculation is not accurate. This is an unweighted average. The actual width requires a weighted average. I have one batch with a row of 1 MB in size, and another batch of 1K rows of 1K each. The average row width is actually ~2K, not ~500K.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/AbstractRecordBatchMemoryManager.java:
```diff
@@ -29,6 +29,50 @@
+  }
+
+  public long getAvgInputRowWidth() {
+    return RecordBatchSizer.safeDivide(sumInputRowWidths, numIncomingBatches);
```
See below regarding the average calculation.

ppadma commented:
@paul-rogers thanks for the review. Please take a look at updated changes.

ppadma commented:
@paul-rogers I added metrics for merge join also. I refactored AbstractRecordBatchMemoryManager to handle batches from multiple streams. Please review when you get a chance.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java:
```diff
@@ -99,7 +100,23 @@ private void clear() {
+    OUTPUT_BATCH_COUNT,
+    AVG_OUTPUT_BATCH_BYTES,
+    AVG_OUTPUT_ROW_BYTES,
+    TOTAL_OUTPUT_RECORDS;
```
Very nice, these labels should be quite readable in the UI. (We should have an enum-to-label lookup table, but we don't...)

This pass I noticed the inconsistency between `INPUT_BATCH_COUNT` and `TOTAL_INPUT_RECORDS`. Perhaps change the latter to `INPUT_RECORD_COUNT`.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/flatten/FlattenRecordBatch.java:
```diff
@@ -463,4 +489,21 @@ protected boolean setupNewSchema() throws SchemaChangeException {
+    stats.setLongStat(Metric.OUTPUT_BATCH_COUNT, flattenMemoryManager.getNumOutgoingBatches());
+    stats.setLongStat(Metric.AVG_OUTPUT_BATCH_BYTES, flattenMemoryManager.getAvgOutputBatchSize());
+    stats.setLongStat(Metric.AVG_OUTPUT_ROW_BYTES, flattenMemoryManager.getAvgOutputRowWidth());
+    stats.setLongStat(Metric.TOTAL_OUTPUT_RECORDS, flattenMemoryManager.getTotalOutputRecords());
```
Not critical, but I found it super helpful to dump this information into the log as well as as metrics. That way, when some issue arises, all the information is in the logs, I didn't need to track down both the logs and the query profile.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/MergeJoinBatch.java:
```diff
@@ -109,12 +110,37 @@
+    LEFT_INPUT_BATCH_COUNT,
+    LEFT_AVG_INPUT_BATCH_BYTES,
+    LEFT_AVG_INPUT_ROW_BYTES,
+    LEFT_TOTAL_INPUT_RECORDS,
```
`LEFT_INPUT_RECORD_COUNT`?

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/MergeJoinBatch.java:
```diff
@@ -127,17 +153,20 @@
+        case LEFT_INDEX:
+          setRecordBatchSizer(inputIndex, new RecordBatchSizer(left));
+          leftRowWidth = getRecordBatchSizer(inputIndex).netRowWidth();
+          logger.debug("mergejoin left incoming batch sizer : {}", getRecordBatchSizer(inputIndex));
```
Maybe "Left incoming batch size"? (The log will tell you it is the merge join. The "sizer" is an implementation detail.)

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/MergeJoinBatch.java:
```diff
@@ -153,16 +182,22 @@ public void update(int inputIndex) {
+      status.setTargetOutputRowCount(adjustOutputRowCount(status.getOutPosition() + numOutputRowsRemaining));
       setOutgoingRowWidth(newOutgoingRowWidth);
+
+      logger.debug("outputBatchSize : {}, avgOutgoingRowWidth : {}, outputRowCount : {}",
```
Since this is a label, not a symbol, maybe "output batch size: {}"

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/MergeJoinBatch.java:
```diff
@@ -573,4 +610,21 @@ private LogicalExpression materializeExpression(LogicalExpression expression, It
+    stats.setLongStat(MergeJoinBatch.Metric.OUTPUT_BATCH_COUNT, mergeJoinMemoryManager.getNumOutgoingBatches());
+    stats.setLongStat(MergeJoinBatch.Metric.AVG_OUTPUT_BATCH_BYTES, mergeJoinMemoryManager.getAvgOutputBatchSize());
+    stats.setLongStat(MergeJoinBatch.Metric.AVG_OUTPUT_ROW_BYTES, mergeJoinMemoryManager.getAvgOutputRowWidth());
+    stats.setLongStat(MergeJoinBatch.Metric.TOTAL_OUTPUT_RECORDS, mergeJoinMemoryManager.getTotalOutputRecords());
```
As above, maybe also write the information to the log for easier debugging.

ppadma commented:
@paul-rogers thanks a lot for the review. updated PR with review comments taken care of.
